### PR TITLE
Switch from go/flags to spf13/cobra

### LINKED ISSUE #7: Switch from go/flags to Cobra
Move from the Golang flags package for the CLI options to Cobra.

Example/more details:

https://blog.alexellis.io/5-keys-to-a-killer-go-cli/

openfaas/faas-cli is a good sample to follow, or alexellis/jaas which is much smaller.

-server=true/false would become a sub-command along with client

```
inlets server --port
inlets client --port --remote / -r
```

The changes should also cover the README.md / instructions.

Use `git commit -s` for commits.

### PR DESCRIPTION
**Pull request by RaviTezu**
Signed-off-by: RaviTezu <ravi-teja@live.com>

## Description/Changes
This PR replaces the go/flags package with spf13/cobra for parsing the command line arguments: 
- Moved parse_upstream.go code into the cmd/client.go as we are do the parsing only in the client mode. Thoughts? 
- Introduced pkg/errors to wrap the context around the errors which are retuned. Thoughts? We can just log/Print them as we are currently doing, no strong feelings here. But I think, wrapping errors using pkg/errors will help in future as the codebase bigger and bigger.
- Updated the README.md file with the changes introduced to the commands. 

## How Has This Been Tested?
Looks like we are yet to add the unit tests, so I have tested the changes by manually running the commands and see how the arguments are being parsed in client and server modes. 

## How are existing users impacted? What migration steps/scripts do we need?
Yes, there's a change in the way the existing users run the commands. For example, `-server=true/false` has been split into two subcommands like `inlets server` and `inlets client`. I have updated the README.md with the changes. I think, we will need to bump up the version? Feel free to let me know I will update this PR with bumping up the version. 

## Checklist:
I have:
- [X] updated the documentation and/or roadmap (if required)
- [X] read the [CONTRIBUTION](https://github.com/alexellis/inlets/blob/master/CONTRIBUTING.md) guide
- [X] signed-off my commits with `git commit -s`

Adding [WIP] to the PR title as there are some open questions and will remove it once we have the answers. 
Closes #7

### CODE REVIEW

**Comment by RaviTezu** [comment]
@alexellis I see the build is failing as I have moved the code in parse_upstream.go to cmd/client.go. I will wait for the your reply before making any changes to the travis config file.

**Comment by alexellis** [comment]
Why does the build fail? Can you fix it by editing `Dockerfile`?

**Comment by RaviTezu** [comment]
@alexellis Yes, fixed it by editing the Docker file.

File: cmd/client.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+	return items
+}
+
+// clientCmd represents the client sub command.
+var clientCmd = &cobra.Command{
+	Use:   "client",
+	Short: "Start the tunnel client.",
+	Long: `Start the tunnel client.
+
+Example: inlets client -remote=192.168.0.101:80 -upstream=http://127.0.0.1:3000
```
Comment (alexellis) [inline cmd/client.go]: This has the old syntax, should be `--`, not `-`

File: cmd/server.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+	if err != nil {
+		return errors.Wrap(err, "failed to get 'token' value.")
+	}
+
+	printToken, err := cmd.Flags().GetBool("print-token")
+	if err != nil {
+		return errors.Wrap(err, "failed to get 'print-token' value.")
+	}
+
+	if len(token) >0 && printToken{
```
Comment (alexellis) [inline cmd/server.go]: You should run `gofmt -w -s` on these files and rebase the commit.

File: cmd/server.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+		Port: port,
+		GatewayTimeout: gatewayTimeout,
+		Token: token,
+	}
+
+	inletsServer.Serve()
+	return nil
+}
+
+func init() {
```
Comment (alexellis) [inline cmd/server.go]: Move to top of file?

File: README.md [hunk 4]
```diff
@@ -188,12 +188,12 @@ spec:
         imagePullPolicy: Always
         command: ["./inlets"]
         args:
-        - "-server=false"
-        - "-upstream=http://gateway.openfaas:8080"
-        - "-remote=your-public-ip"
+        - "client"
+        - "--upstream=http://gateway.openfaas:8080"
```
Comment (alexellis) [inline README.md]: How about giving an example with multiple `--upstream`?

**Comment by alexellis** [comment]
FYI we now have a `-version` command to add as `inlets version` after merging #17 

Alex

**Comment by RaviTezu** [comment]
@alexellis Included the version changes as well.

File: README.md [hunk 5]
```diff
@@ -108,19 +108,20 @@ port=3000 go run server.go
...
-Replace the `-remote` with the address where your other machine is listening.
+Replace the `--remote` with the address where your other machine is listening.
 
 We now have an example service running (hash-browns), a tunnel server and a tunnel client.
 
 So send a request to the public IP address or hostname:
 
-```sh
-./inlets -server=false -remote=192.168.0.101:80 -upstream  "gateway.mydomain.tk=http://127.0.0.1:3000"
+<<<<<<< HEAD
```
Comment (alexellis) [inline README.md]: Looks like the merge went wrong. Please can you fix? ^

File: README.md [hunk 6]
```diff
@@ -200,12 +201,12 @@ spec:
         imagePullPolicy: Always
         command: ["./inlets"]
         args:
-        - "-server=false"
-        - "-upstream=http://gateway.openfaas:8080"
-        - "-remote=your-public-ip"
+        - "client"
+        - "--upstream=http://gateway.openfaas:8080,http://endpoint.openfaas:9090"
```
Comment (alexellis) [inline README.md]: Maybe it should be in this PR or a new PR - to be able to specify `--upstream` multiple times:

```
--upstream=http://gateway.openfaas:8080
--upstream=http://endpoint.openfaas:9090
```

The way I implemented it before was to work around go/flags limitations.

File: cmd/client.go [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+	"log"
+	"strings"
+
+	"github.com/alexellis/inlets/pkg/client"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+)
+
+func init() {
+	InletsCmd.AddCommand(clientCmd)
```
Comment (alexellis) [inline cmd/client.go]: Should `InletsCmd` be `inletsCmd` ?

File: cmd/client.go [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+import (
+	"log"
+	"strings"
+
+	"github.com/alexellis/inlets/pkg/client"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+)
+
+func init() {
```
Comment (alexellis) [inline cmd/client.go]: Is there an `inlets help / inlets --help` command for the root?

File: README.md (hunk 5 above)
Comment (RaviTezu) [reply README.md]: ah, removed it.

File: cmd/client.go (hunk 7 above)
Comment (RaviTezu) [reply cmd/client.go]: changed it to `inletsCmd`

File: cmd/client.go (hunk 8 above)
Comment (RaviTezu) [reply cmd/client.go]: Yes, it does provide `-h/--help`. Something like this: 
```sh
$> ./inlets --help                                                                                                                                                                                                                                               
Inlets combines a reverse proxy and websocket tunnels to expose your internal and development
endpoints to the public Internet via an exit-node.

An exit-node may be a 5-10 USD VPS or any other computer with an IPv4 IP address.

See: https://github.com/alexellis/inlets for more information.

Usage:
  inlets [flags]
  inlets [command]

Available Commands:
  client      Start the tunnel client.
  help        Help about any command
  server      Start the tunnel server on a machine with a publicly-accessible IPv4 IP address such as a VPS.

Flags:
  -h, --help      help for inlets
  -v, --version   print the version information

Use "inlets [command] --help" for more information about a command.
```

File: README.md (hunk 6 above)
Comment (RaviTezu) [reply README.md]: I just check the cobra documentation and looks like it is not straight forward. Let's target this in another PR.

**Comment by RaviTezu** [comment]
@alexellis Thanks for the review. Pushed the changes.

**Comment by alexellis** [comment]
Awesome, I'll give it another pass soon and hopefully get this merged/released.

**Comment by alexellis** [comment]
Have we also updated the service entries for systemd?

**Comment by RaviTezu** [comment]
@alexellis Updated the systemd file.

**Comment by alexellis** [comment]
We have some conflicts with the base branch which means I cannot rebase your PR into master.

Please do the following to sync up with master and also to collapse all of your commits into a single commit, removing the accidental merge at the same time:

```
git checkout replace_flags_with_cobra
git checkout -b replace_flags_with_cobra_backup # make a backup first

git remote upstream add github.com/alexellis/inlets
git checkout replace_flags_with_cobra
git reset HEAD~4 # where 4 is your number of changes / commits
git add .
git stash # temporarily stash your changes

git merge upstream/master --rebase # important to use rebase here
git stash pop # get your changes back
git add .
git commit -s # now add your message for the whole set of changes
git push origin replace_flags_with_cobra --force # sync the PR
```

Thank you, looking forward to merging this.

**Comment by alexellis** [comment]
@RaviTezu FYI

**Comment by RaviTezu** [comment]
@alexellis Done

**Comment by RaviTezu** [comment]
@alexellis I just did a `docker build .` in `~/go/src/github.com/alexellis/inlets` and it worked. Can we trigger the check again?

**Comment by alexellis** [comment]
This is the error

```

12.49s$ make dist
CGO_ENABLED=0 GOOS=linux go build -ldflags "-s -w -X main.Version=0.5.7-7-g43d7c62 -X main.GitCommit=43d7c62b767746a0de5bb3e250582ac21fde78f3" -a -installsuffix cgo -o bin/inlets
# github.com/alexellis/inlets/cmd
cmd/client.go:13:2: undefined: inletsCmd
cmd/server.go:13:2: undefined: inletsCmd
make: *** [dist] Error 2
The command "make dist" exited with 2.
```

You can view the logs on Travis following the status link.

**Comment by RaviTezu** [comment]
@alexellis Both the commands are working fine on my local machine. Something wrong with travis config?
```
╰─ make docker                                                                                                                                                                                                                                                                                                 
docker build --build-arg Version=0.5.7-6-ge66d73f --build-arg GIT_COMMIT=e66d73f50d1e41af0d966d0409ef5906bf507959 -t alexellis2/inlets:0.5.7-6-ge66d73f .
Sending build context to Docker daemon   30.7MB
Step 1/19 : FROM golang:1.10 as build
 ---> 6fd1f7edb6ab
Step 2/19 : WORKDIR /go/src/github.com/alexellis/inlets
 ---> Using cache
 ---> ec434f8df965
Step 3/19 : COPY .git               .git
 ---> Using cache
 ---> 0d88ba4f481e
Step 4/19 : COPY vendor             vendor
 ---> Using cache
 ---> 81355bbfc2d0
Step 5/19 : COPY pkg                pkg
 ---> Using cache
 ---> 16a07b7d4639
Step 6/19 : COPY cmd                cmd
 ---> Using cache
 ---> e478f0cacb2a
Step 7/19 : COPY main.go            .
 ---> Using cache
 ---> d1876458f660
Step 8/19 : ARG GIT_COMMIT
 ---> Using cache
 ---> 9a0d71a54a99
Step 9/19 : ARG VERSION
 ---> Using cache
 ---> ca89ee0635de
Step 10/19 : RUN CGO_ENABLED=0 go build -ldflags "-s -w -X main.GitCommit=${GIT_COMMIT} -X main.Version=${VERSION}" -a -installsuffix cgo -o /usr/bin/inlets
 ---> Running in 0ab8a7da5006
Removing intermediate container 0ab8a7da5006
 ---> 7ed8b1a7ddea
Step 11/19 : FROM alpine:3.9
 ---> 5cb3aa00f899
Step 12/19 : RUN apk add --force-refresh ca-certificates
 ---> Using cache
 ---> 295697f25b93
Step 13/19 : RUN addgroup -S app && adduser -S -g app app   && mkdir -p /home/app || :   && chown -R app /home/app
 ---> Using cache
 ---> bd4ca52535d9
Step 14/19 : COPY --from=build /usr/bin/inlets /usr/bin/
 ---> 94fa1e44772f
Step 15/19 : WORKDIR /home/app
 ---> Running in 38062e00d2ac
Removing intermediate container 38062e00d2ac
 ---> a5cf465fd776
Step 16/19 : USER app
 ---> Running in 6fe99b17ddde
Removing intermediate container 6fe99b17ddde
 ---> 1aa4ba3442d0
Step 17/19 : EXPOSE 80
 ---> Running in e1d7d8ad61c9
Removing intermediate container e1d7d8ad61c9
 ---> 2981f83062e9
Step 18/19 : ENTRYPOINT ["inlets"]
 ---> Running in 769e00bf0f9a
Removing intermediate container 769e00bf0f9a
 ---> 1473b1e04aa3
Step 19/19 : CMD ["-help"]
 ---> Running in a00d5ee27bda
Removing intermediate container a00d5ee27bda
 ---> 49fa109be872
[Warning] One or more build-args [Version] were not consumed
Successfully built 49fa109be872
Successfully tagged alexellis2/inlets:0.5.7-6-ge66d73f

╰─ make dist                                                                                                                                                                                                                                                                                                  
CGO_ENABLED=0 GOOS=linux go build -ldflags "-s -w -X main.Version=0.5.7-6-ge66d73f -X main.GitCommit=e66d73f50d1e41af0d966d0409ef5906bf507959" -a -installsuffix cgo -o bin/inlets
CGO_ENABLED=0 GOOS=darwin go build -ldflags "-s -w -X main.Version=0.5.7-6-ge66d73f -X main.GitCommit=e66d73f50d1e41af0d966d0409ef5906bf507959" -a -installsuffix cgo -o bin/inlets-darwin
CGO_ENABLED=0 GOOS=linux GOARCH=arm GOARM=6 go build -ldflags "-s -w -X main.Version=0.5.7-6-ge66d73f -X main.GitCommit=e66d73f50d1e41af0d966d0409ef5906bf507959" -a -installsuffix cgo -o bin/inlets-armhf
CGO_ENABLED=0 GOOS=linux GOARCH=arm64 go build -ldflags "-s -w -X main.Version=0.5.7-6-ge66d73f -X main.GitCommit=e66d73f50d1e41af0d966d0409ef5906bf507959" -a -installsuffix cgo -o bin/inlets-arm64
```

**Comment by RaviTezu** [comment]
@alexellis I see the issue. `.gitignore` is causing the issues. Add an extra check to allow `inlets.go` to consider the `intlets.go` file. Build should work fine now.

**Comment by alexellis** [comment]
Thank you so much, this PR looks 💯 

There's one more that I think is important for us to look at next: #42 

Also remind me when we have a logo & a sticker to post you one 👍

**Pull request merged**